### PR TITLE
Laboratory work #1, Kondrashova Ekaterina - 20FPL1

### PR DESCRIPTION
**Pull request by eskondrashova**

### CODE REVIEW

**Comment by ValeriyaKuznetsova** [comment]
It's time to push some changes. October 1st is coming

File: lab_1/main.py [hunk 1]
```diff
@@ -98,22 +195,63 @@ def detect_language_advanced(unknown_profile: dict, profiles: list, languages: l
...
 
-def load_profile(path_to_file: str) -> dict or None:
+def load_profile(path_to_file):
     """
     Loads a language profile
     :param path_to_file: a path
     :return: a dictionary with three keys – name, freq, n_words
     """
-    pass
+    if isinstance(path_to_file, str):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:231]: check

**Review by ValeriyaKuznetsova — COMMENTED** [review]
In general, it's nice. I have some comments on style and the implementation, as well as some questions. These questions are obligatory to answer – here or via Telegram

File: lab_1/main.py [hunk 2]
```diff
@@ -2,70 +2,127 @@
 Lab 1
 Language detection
 """
+import json
+from os.path import exists
 
 
-def tokenize(text: str) -> list or None:
+def tokenize(text):
```
[the remark this answers is not in the document]
Comment (eskondrashova) [reply lab_1/main.py:9]: done

File: lab_1/main.py [hunk 3]
```diff
@@ -2,70 +2,127 @@
...
     """
     Splits a text into tokens, converts the tokens into lowercase,
     removes punctuation and other symbols from words
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
     """
-    pass
+    if isinstance(text, str):
```
[the remark this answers is not in the document]
Comment (eskondrashova) [reply lab_1/main.py]: done

File: lab_1/main.py [hunk 4]
```diff
@@ -2,70 +2,127 @@
...
-def calculate_frequencies(tokens: list) -> dict or None:
+def calculate_frequencies(tokens):
     """
     Calculates frequencies of given tokens
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
-    pass
+    freq_dict = {}
+    if isinstance(tokens, list) and all(isinstance(s, str) for s in tokens):
```
[the remark this answers is not in the document]
Comment (eskondrashova) [reply lab_1/main.py]: done

File: lab_1/main.py [hunk 5]
```diff
@@ -2,70 +2,127 @@
...
-def get_top_n_words(freq_dict: dict, top_n: int) -> list or None:
+def get_top_n_words(freq_dict, top_n):
     """
     Returns the most common words
     :param freq_dict: a dictionary with frequencies
     :param top_n: a number of the most common words
     :return: a list of the most common words
     """
-    pass
+    if not isinstance(freq_dict, dict):
```
[the remark this answers is not in the document]
Comment (eskondrashova) [reply lab_1/main.py]: done

File: lab_1/main.py [hunk 6]
```diff
@@ -2,70 +2,127 @@
...
+        return None
+    if not all(isinstance(v, int) for v in freq_dict.values()):
+        return None
+    if not isinstance(top_n, int):
+        return None
+    srtd_list = [wrd[0] for wrd in sorted(freq_dict.items(), key=lambda val: val[1], reverse=True)]
```
[the remark this answers is not in the document]
Comment (eskondrashova) [reply lab_1/main.py]: done

File: lab_1/main.py [hunk 7]
```diff
@@ -2,70 +2,127 @@
...
+def compare_profiles(unknown_profile, profile_to_compare, top_n):
     """
     Compares profiles and calculates the distance using top n words
     :param unknown_profile: a dictionary
     :param profile_to_compare: a dictionary
     :param top_n: a number of the most common words
     :return: the distance
     """
-    pass
+    if (
```
[the remark this answers is not in the document]
Comment (eskondrashova) [reply lab_1/main.py]: redone

File: lab_1/main.py [hunk 8]
```diff
@@ -2,70 +2,127 @@
...
+        isinstance(unknown_profile, dict)
+        and isinstance(profile_to_compare, dict)
+        and isinstance(top_n, int)
+    ):
+        freq_list1 = unknown_profile['freq']
+        top_n1 = get_top_n_words(freq_list1, top_n)
```
[the remark this answers is not in the document]
Comment (eskondrashova) [reply lab_1/main.py]: done

File: lab_1/main.py [hunk 9]
```diff
@@ -2,70 +2,127 @@
...
+        freq_list2 = profile_to_compare['freq']
+        top_n2 = get_top_n_words(freq_list2, top_n)
 
+        profiles_in_common = []
+        for wrd in top_n1:
+            if wrd in top_n2:
```
[the remark this answers is not in the document]
Comment (eskondrashova) [reply lab_1/main.py]: done

File: lab_1/main.py [hunk 10]
```diff
@@ -74,10 +131,23 @@ def detect_language(unknown_profile: dict, profile_1: dict, profile_2: dict, top
...
+        if compare_1 > compare_2:
+            return profile_1['name']
+        if compare_2 > compare_1:
+            return profile_2['name']
+        return max(profile_1['name'], profile_2['name'])
+    return None
 
 
-def compare_profiles_advanced(unknown_profile: dict, profile_to_compare: dict, top_n: int) -> list or None:
+def compare_profiles_advanced(unknown_profile, profile_to_compare, top_n):
```
[the remark this answers is not in the document]
Comment (eskondrashova) [reply lab_1/main.py]: done

File: lab_1/main.py [hunk 11]
```diff
@@ -98,22 +195,63 @@ def detect_language_advanced(unknown_profile: dict, profiles: list, languages: l
...
+        dict_lang_score = {}
+        for profile_to_compare in profiles:
+            if languages == []:
+                compare = compare_profiles_advanced(unknown_profile, profile_to_compare, top_n)
+                dict_lang_score[compare['name']] = compare['score']
+            else:
+                if profile_to_compare['name'] in languages:
+                    compare = compare_profiles_advanced(unknown_profile, profile_to_compare, top_n)
+                    dict_lang_score[compare['name']] = compare['score']
+        if dict_lang_score != {}:
```
[the remark this answers is not in the document]
Comment (eskondrashova) [reply lab_1/main.py]: done

File: lab_1/main.py [hunk 12]
```diff
@@ -98,22 +195,63 @@ def detect_language_advanced(unknown_profile: dict, profiles: list, languages: l
...
-    pass
+    if (
+        isinstance(unknown_profile, dict)
+        and isinstance(profiles, list)
+        and isinstance(languages, list)
+        and isinstance(top_n, int)
+    ):
+        dict_lang_score = {}
+        for profile_to_compare in profiles:
+            if languages == []:
```
[the remark this answers is not in the document]
Comment (eskondrashova) [reply lab_1/main.py]: done

File: lab_1/main.py [hunk 13]
```diff
@@ -6,123 +6,135 @@
...
-                return cleaned_tokens
-            return tokens
+    if not (
+        isinstance(tokens, list)
+        and tokens != []
+        and all(isinstance(s, str) for s in tokens)
+    ):
         return None
-    return None
+    if isinstance(stop_words, list):
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:38]: why do you need this check?

File: lab_1/main.py [hunk 14]
```diff
@@ -2,70 +2,127 @@
...
+def remove_stop_words(tokens, stop_words):
     """
     Removes stop words
     :param tokens: a list of tokens
     :param stop_words: a list of stop words
     :return: a list of tokens without stop words
     """
-    pass
+    if isinstance(tokens, list) and tokens != []:
+        if all(isinstance(s, str) for s in tokens):
```
[the remark this answers is not in the document]
Comment (ValeriyaKuznetsova) [reply lab_1/main.py]: not answered

File: lab_1/main.py [hunk 15]
```diff
@@ -2,70 +2,127 @@
...
     """
     Returns the most common words
     :param freq_dict: a dictionary with frequencies
     :param top_n: a number of the most common words
     :return: a list of the most common words
     """
-    pass
+    if not isinstance(freq_dict, dict):
+        return None
+    if not all(isinstance(v, int) for v in freq_dict.values()):
```
[the remark this answers is not in the document]
Comment (ValeriyaKuznetsova) [reply lab_1/main.py]: not answered

File: lab_1/main.py [hunk 16]
```diff
@@ -131,23 +143,25 @@ def detect_language(unknown_profile, profile_1, profile_2, top_n):
...
-        compare_1 = compare_profiles(unknown_profile, profile_1, top_n)
-        compare_2 = compare_profiles(unknown_profile, profile_2, top_n)
-        if compare_1 > compare_2:
-            return profile_1['name']
-        if compare_2 > compare_1:
-            return profile_2['name']
-        return max(profile_1['name'], profile_2['name'])
-    return None
+        return None
+    compare_1 = compare_profiles(unknown_profile, profile_1, top_n)
```
Comment (ValeriyaKuznetsova) [inline lab_1/main.py:153]: the same about compare_1 -> first_compare

don't use numbers in names

File: lab_1/main.py [hunk 17]
```diff
@@ -98,22 +195,63 @@ def detect_language_advanced(unknown_profile: dict, profiles: list, languages: l
...
-    pass
+    if not isinstance(profile, dict):
+        return 1
+    if not (
+        isinstance(profile['name'], str)
+        or isinstance(profile['freq'], dict)
+        or isinstance(profile['n_words'], int)
+    ):
+        return 1
+    profile_dict = '{}.json'.format(profile['name'])
```
[the remark this answers is not in the document]
Comment (ValeriyaKuznetsova) [reply lab_1/main.py:266]: not answered

**Review by ValeriyaKuznetsova — COMMENTED** [review]
I reviewed it. Two minor comments

By the way, you didn't answer my questions. They're obligatory to answer. I won't put `Done` until I see your answers

**Comment by ValeriyaKuznetsova** [comment]
Good job! Your code looks great. Keep it up!

**Pull request closed without merging**